File: Javascript Codewars/6kyu/Exclamation marks series #12: Remove odd number continuous exclamation marks and question marks.js
/*
Description

Remove odd number continuous exclamation marks and question marks(from the left to the right), until no continuous 
exclamation marks and question marks exist. Please note: One exclamation mark or question mark is not a continuous 
exclamation marks or question marks. The string only contains ! and ?.

Examples
remove("") === ""
remove("!") === "!"
remove("!!") === "!!"
remove("!!!") === ""
remove("!??") === "!??"
remove("!???") === "!"
remove("!!!??") === "??"
remove("!!!???") === ""
remove("!??") === "!??"
remove("!???!!") === ""
("!???!!" --> "!!!" --> "")
remove("!????!!!?") === "!"
("!????!!!?" --> "!?????" --> "!")

Note
Please don't post issue about difficulty or duplicate.
*/

function remove(s) {
  while ((s.match(/(.)\1{2,}/g) || []).some((l) => l.length % 2)) {
    s = s
      .match(/(.)\1*/g)
      .filter((x) => x.length < 3 || !(x.length % 2))
      .join("");
  }
  return s;
}

File: Python Codewars/6kyu/Exclamation marks series #12: Remove odd number continuous exclamation marks and question marks.py
'''
Description

Write a function that removes all odd number continuous exclamation marks and question marks in a string, 
repeat this operation until no continuous exclamation marks and question marks of odd length exist.

Please note: One exclamation mark or question mark is not continuous exclamation marks or question marks.

The string contains only ! and ?.

Examples
"" --> ""
"!" --> "!"
"!!" --> "!!"
"!!!" --> ""
"!??" --> "!??"
"!???" --> "!"
"!!!??" --> "??"
"!!!???" --> ""
"!???!!" --> "!!!" --> ""
"!????!!!?" --> "!?????" --> "!"
'''

import re

def remove(inp):
    while True:
        next = re.sub(r'!{3,}|\?{3,}', lambda m: "" if len(m.group(0)) % 2 == 1 else m.group(0), inp)
        if next == inp: break
        inp = next
    return next